File: hora/horoscope/chart/ashtakavarga.py
import numpy as np
from hora import const, utils

planet_list = ['sun','moon','mars','mercury','jupiter','venus','saturn','lagnam']
raasi_list=['Mesham','Rishabam','Mithunam','Katakam','Simmam','Kanni','Thulaam','Vrichigam','Dhanusu','Makaram','Kumbam','Meenam']
raasi_index = lambda planet,planet_positions_in_chart: [i for i,raasi in enumerate(planet_positions_in_chart) if planet !=const._ascendant_symbol and planet.lower() in raasi.lower() ][0]
def get_ashtaka_varga(house_to_planet_list):
    """
        get binna, samudhaya and prastara varga from the given horoscope chart
        @param house_to_planet_list: 1-D array [0..11] with planets in each raasi
            Example: ['','','','','2','7','1/5','0','3/4','L','','6/8']
        @return: 
            binna_ashtaka_varga - 2-D List [0..7][0..7] 0=Sun..7=Lagnam
            samudhaya ashtaka varga - 1D List [0..11] 0=Aries 11=Pisces
            prastara ashtaka varga - 3D List [0..7][0..7][0..11]
    """
    #print('get_ashtaka_varga','house_to_planet_list',house_to_planet_list)
    p_to_h = utils.get_planet_to_house_dict_from_chart(house_to_planet_list)
    #print('get_ashtaka_varga','p_to_h',p_to_h)
    raasi_ashtaka = [[0 for r in range(12)] for p in range(8)]
    prastara_ashtaka_varga  = [[[0 for r in range(12)] for p1 in range(10)] for p2 in range(8)]
    for key in const.ashtaka_varga_dict.keys():
        p = int(key)
        #planet = planet_list[p]
        planet_raasi_list = const.ashtaka_varga_dict[key]
        for op,other_planet in enumerate(planet_raasi_list):
            pr = p_to_h[op]
            if op == 7: #Lagnam
                pr = p_to_h[const._ascendant_symbol]
            for raasi in other_planet:
                r = (raasi-1+pr) % 12
                raasi_ashtaka[p][r] +=1
                prastara_ashtaka_varga[p][op][r] = 1
                prastara_ashtaka_varga[p][-1][r] += 1
    binna_ashtaka_varga = raasi_ashtaka[0:8][:]
    prastara_ashtaka_varga = prastara_ashtaka_varga[0:8][0:9][:]
    samudhaya_ashtaka_varga = np.asarray(binna_ashtaka_varga[:-1]).sum(axis=0).tolist() # [0:-1] to exlcude Lagnam
    return binna_ashtaka_varga, samudhaya_ashtaka_varga,prastara_ashtaka_varga
def _trikona_sodhana(binna_ashtaka_varga):
    bav = binna_ashtaka_varga[:]
    for p in range(7):
        for r in range(4):
            if bav[p][r]==0 or bav[p][r+4]==0 or bav[p][r+8]==0:
                #print('Rule 1:If atleast one rasi has zero, no reduction is necessary.',p,r,bav[p][r],bav[p][r+4],bav[p][r+8])
                continue
            elif bav[p][r]==bav[p][r+4] and bav[p][r+4]==bav[p][r+8]:
                #print('Rule (2) If the three rasis have the same value, make them all zero.',p,r,bav[p][r],bav[p][r+4],bav[p][r+8])
                bav[p][r]=0
                bav[p][r+4]=0
                bav[p][r+8]=0
            else:
                #print('Rule (3) Take the lowest value out of the three. Subtract it from all the values.',p,r,bav[p][r],bav[p][r+4],bav[p][r+8])
                min_value = min([bav[p][r],bav[p][r+4],bav[p][r+8]])
                #print('before',p,r,min_value,bav[p][r],bav[p][r+4],bav[p][r+8])
                bav[p][r] -= min_value
                bav[p][r+4] -= min_value
                bav[p][r+8] -= min_value
                #print('after',p,r,min_value,bav[p][r],bav[p][r+4],bav[p][r+8])
    return bav
def _ekadhipatya_sodhana(binna_ashtaka_varga_after_trikona,chart_1d):
    bav = binna_ashtaka_varga_after_trikona[:]
    rasi_owners=[4,3,(0,7),(2,5),(8,11),(1,6),(9,10)]
    for p in range(2,7):
        r1,r2 = rasi_owners[p]
        r1_occupied = not (chart_1d[r1].strip() == '')
        r2_occupied = not (chart_1d[r2].strip() == '')
        # Rule 1 (either bav is 0. Rule 2: either rasi is occupied by a planet
        if (bav[p][r1]==0 or bav[p][r2]==0) or (r1_occupied and r2_occupied) :
            continue
        elif (not r1_occupied) and (not r2_occupied ): # Rule 4 - both rasi are empty
            # Rule 4(b) (b) If they have different values, replace the higher value with the lower value.
            if bav[p][r1] != bav[p][r2]: 
                min_value = min([bav[p][r1],bav[p][r2]])
                bav[p][r1] = min_value
                bav[p][r2] = min_value
            # Rule 4(a) If both the rasis have the same value, replace both with zero.
            else:
                bav[p][r1] = 0
                bav[p][r2] = 0
        else:
            # Rule (3) If one rasi is occupied by a planet (or planets) and the other is empty
            if (r1_occupied): #r2 is empty
                # Rule 3(a) If the empty rasi has a lower value, replace the value with a zero.
                if (bav[p][r2] < bav[p][r1]):
                    bav[p][r2] = 0
                # Rule 3(b) If the empty rasi has a higher value, replace the value with the value in the other rasi.
                else:
                    bav[p][r2] = bav[p][r1]
            else: #r1 is empty
                # Rule 3(a) If the empty rasi has a lower value, replace the value with a zero. 
                if (bav[p][r1] < bav[p][r2]):
                    bav[p][r1] = 0
                # Rule 3(b) If the empty rasi has a higher value, replace the value with the value in the other rasi.
                else:
                    bav[p][r1] = bav[p][r2]
    return bav
def _get_planet_positions(chart_1d):
    planet_houses = [-1 for p in range(7)]
    for p,planet in enumerate(planet_list[0:-1]): # Excluding Lagnam
        for house,rasi in enumerate(chart_1d):
            if planet.lower() in rasi.lower():
                planet_houses[p] = house
                break
    return planet_houses
def _sodhya_pindas(binna_ashtaka_varga_after_ekadhipatya,chart_1d):
    rasimana_multipliers = [7,10,8,4,10,6,7,8,9,5,11,12]
    grahamana_multipliers = [5,5,8,5,10,7,5]
    bav = binna_ashtaka_varga_after_ekadhipatya[:]
    raasi_pindas = [0 for p in range(7)]
    graha_pindas = [0 for p in range(7)]
    sodhya_pindas = [0 for p in range(7)]
    for p in range(7):
        raasi_pindas[p] = sum(np.multiply(bav[p][:],rasimana_multipliers))
    #planet_houses = _get_planet_positions(chart_1d)
    planet_houses = list(utils.get_planet_to_house_dict_from_chart(chart_1d).values())[:7] # Exclude Rahu, Ketu and Lagnam
    for p in range(7):
        graha_pindas[p] = sum([ grahamana_multipliers[i]*bav[p][pr] for i,pr in enumerate(planet_houses)])
        sodhya_pindas[p] = raasi_pindas[p]+graha_pindas[p]
    return raasi_pindas,graha_pindas,sodhya_pindas
def sodhaya_pindas(binna_ashtaka_varga,house_to_planet_chart):
    """
        Get sodhaya pindas from binna ashtaka varga
        @param param:binna_ashtaka_varga - 2-D List [0..7][0..7] 0=Sun..7=Lagnam - of BAV values
        NOTE: To pass binn ashtaka varga as parameter - you need to get it from get_ashtaka_varga function
        @return: raasi_pindas,graha_pindas,sodhya_pindas
                raasi_pindas : raasi pindas of planets 0=Sun to 6=Saturn [0..6]
                graha_pindas : graha pindas of planets 0=Sun to 6=Saturn [0..6]
                sidhaya_pindas : sodhaya pindas of planets 0=Sun to 6=Saturn [0..6]
    """
    #print('bav before trikona',binna_ashtaka_varga)
    binna_ashtaka_varga_after_trikona = _trikona_sodhana(binna_ashtaka_varga)
    #print('bav after trikona',binna_ashtaka_varga_after_trikona)
    binna_ashtaka_varga_after_ekadhipatya = _ekadhipatya_sodhana(binna_ashtaka_varga_after_trikona,house_to_planet_chart)
    """ binna_ashtaka_varga_after_ekadhipatya is called Sodhita Ashtakavarga"""
    #print('Sodhita Ashtakavarga\n',binna_ashtaka_varga_after_ekadhipatya)
    raasi_pindas,graha_pindas,sodhya_pindas = _sodhya_pindas(binna_ashtaka_varga_after_ekadhipatya,house_to_planet_chart)
    return raasi_pindas,graha_pindas,sodhya_pindas
if __name__ == "__main__":
    from hora.tests.pvr_tests import test_example
    # Chart 7 from the book
    chapter = 'Chaper 12.3 ashtaka_varga_tests'
    exercise = 'Exercise 22/Chart 7:'
    chart_7 = ['6/1/7','','','','','','8/4','L','3/2','0','5','']
    bav, sav,pav = get_ashtaka_varga(chart_7)
    bav_e = [[4,2,3,4,6,5,5,3,2,6,6,2],
             [6,3,5,3,5,5,6,3,3,4,4,2],
             [3,2,3,4,2,5,4,3,3,4,3,3],
             [4,6,4,3,4,7,4,5,6,3,5,3],
             [4,4,3,5,6,5,6,4,6,4,3,6],
             [3,5,5,4,6,2,3,6,5,2,7,4],
             [3,2,2,3,5,6,3,4,1,3,6,1]]
    test_example(chapter+exercise+' BAV',bav_e,bav[:-1])#,assert_result=True)
    sav_e = [27,24,25,26,34,35,31,28,26,26,34,21]
    test_example(chapter+exercise+' SAV',sav_e,sav)#,assert_result=True)
    sp_e_book = [[152,85,52,95,68,154,162],[81,55,43,33,56,54,63],[233,140,95,128,124,208,225]]
    sp_e = ([155, 92, 55, 99, 93, 154, 166], [81, 55, 43, 33, 56, 54, 63], [236, 147, 98, 132, 149, 208, 229])
    sp = sodhaya_pindas(bav, chart_7)
    test_example(chapter+exercise+' Sodhaya Pindas',sp_e,sp)
    print(chapter+exercise+' Sodhaya Pindas:\n NOTE: Not clear why this case SP failed to match the book\n'+
          ' Examples 40,41 & 42 based on Chart 12 are matching BAV, SAV and SP.\n So the calculations in this code is thus verified\n'+
          'Expected Values from Book:',sp_e_book)

File: pyjhora-web/src/core/horoscope/ashtakavarga.ts
/**
 * Ashtakavarga Calculations
 * Ported from PyJHora ashtakavarga.py
 *
 * Includes:
 * - Binna Ashtakavarga (BAV) for 7 planets + Lagna
 * - Sarva Ashtakavarga (SAV) - combined totals
 * - Prastara Ashtakavarga - detailed contribution breakdown
 * - Trikona Sodhana - trine reduction
 * - Ekadhipatya Sodhana - dual lordship reduction
 * - Sodhya Pindas - weighted totals (Rasi, Graha, Sodhya)
 */

import {
  ASCENDANT_SYMBOL,
  ASHTAKA_VARGA_DICT,
  GRAHAMANA_MULTIPLIERS,
  RASIMANA_MULTIPLIERS
} from '../constants';
import type { HouseChart } from '../types';

/**
 * Ashtakavarga result containing all three components
 */
export interface AshtakavargaResult {
  /** Binna Ashtakavarga: 8x12 array [planet][rasi] of benefic points */
  binnaAshtakavarga: number[][];
  /** Sarva Ashtakavarga: 12-element array [rasi] of total points (excluding Lagna) */
  sarvaAshtakavarga: number[];
  /** Prastara Ashtakavarga: 8x9x12 array [planet][contributor][rasi] */
  prastaraAshtakavarga: number[][][];
}

/**
 * Sodhya Pindas result
 */
export interface SodhyaPindasResult {
  /** Rasi Pindas for each planet (Sun to Saturn) */
  raasiPindas: number[];
  /** Graha Pindas for each planet (Sun to Saturn) */
  grahaPindas: number[];
  /** Sodhya Pindas for each planet (Rasi + Graha) */
  sodhyaPindas: number[];
}

/**
 * Convert house chart (string format) to planet-to-house dictionary
 * @param houseChart - Array of 12 strings like ['0', '1/2', 'L', ...] where numbers are planet IDs
 * @returns Map of planet ID or 'L' to house index (0-11)
 */
export const getPlanetToHouseFromChart = (
  houseChart: HouseChart
): Record<number | string, number> => {
  const pToH: Record<number | string, number> = {};

  houseChart.forEach((planets, house) => {
    if (!planets || planets.trim() === '') return;

    // Split by '/' to handle multiple planets in same house
    const planetList = planets.split('/');

    planetList.forEach(p => {
      const trimmed = p.trim();
      if (trimmed === ASCENDANT_SYMBOL || trimmed === 'L') {
        pToH[ASCENDANT_SYMBOL] = house;
      } else {
        const planetId = parseInt(trimmed);
        if (!isNaN(planetId)) {
          pToH[planetId] = house;
        }
      }
    });
  });

  return pToH;
};

/**
 * Calculate Ashtakavarga (Binna, Sarva, and Prastara)
 *
 * @param houseChart - 1D array [0..11] with planets in each rasi
 *                     Example: ['', '', '', '', '2', '7', '1/5', '0', '3/4', 'L', '', '6/8']
 * @returns AshtakavargaResult containing BAV, SAV, and Prastara
 */
export const getAshtakavarga = (houseChart: HouseChart): AshtakavargaResult => {
  const pToH = getPlanetToHouseFromChart(houseChart);

  // Initialize arrays
  // raasi_ashtaka[planet][rasi] - 8 planets x 12 rasis
  const raasiAshtaka: number[][] = Array.from({ length: 8 }, () =>
    Array(12).fill(0)
  );

  // prastara_ashtaka_varga[planet][contributor][rasi] - 8 planets x 10 contributors x 12 rasis
  // Contributors: 0-6 = Sun to Saturn, 7 = Lagna, 8 = unused, 9 = total row
  const prastaraAshtakavarga: number[][][] = Array.from({ length: 8 }, () =>
    Array.from({ length: 10 }, () => Array(12).fill(0))
  );

  // Calculate Binna and Prastara Ashtakavarga
  for (let p = 0; p < 8; p++) {
    const planetRaasiList = ASHTAKA_VARGA_DICT[p];
    if (!planetRaasiList) continue;

    for (let op = 0; op < 8; op++) {
      // Get the rasi of the contributing planet/lagna
      let pr: number | undefined;
      if (op === 7) {
        // Lagna
        pr = pToH[ASCENDANT_SYMBOL];
      } else {
        pr = pToH[op];
      }

      if (pr === undefined) continue;

      const beneficHouses = planetRaasiList[op];
      if (!beneficHouses) continue;

      for (const house of beneficHouses) {
        // house is 1-indexed in the dict, convert to 0-indexed rasi
        const r = (house - 1 + pr) % 12;
        raasiAshtaka[p][r] += 1;
        prastaraAshtakavarga[p][op][r] = 1;
        prastaraAshtakavarga[p][9][r] += 1; // Total row
      }
    }
  }

  // Binna Ashtakavarga (first 8 rows)
  const binnaAshtakavarga = raasiAshtaka.slice(0, 8).map(row => [...row]);

  // Prastara (first 8 planets, first 9 rows per planet)
  const prastara = prastaraAshtakavarga
    .slice(0, 8)
    .map(planetArr => planetArr.slice(0, 9).map(row => [...row]));

  // Sarva Ashtakavarga - sum of BAV for planets 0-6 (excluding Lagna at index 7)
  const sarvaAshtakavarga: number[] = Array(12).fill(0);
  for (let r = 0; r < 12; r++) {
    for (let p = 0; p < 7; p++) {
      sarvaAshtakavarga[r] += binnaAshtakavarga[p][r];
    }
  }

  return {
    binnaAshtakavarga,
    sarvaAshtakavarga,
    prastaraAshtakavarga: prastara
  };
};

/**
 * Trikona Sodhana - Trine reduction on Binna Ashtakavarga
 *
 * Rules:
 * 1. If at least one rasi in the trine group has zero, no reduction
 * 2. If all three rasis have the same value, make them all zero
 * 3. Otherwise, subtract the minimum value from all three
 *
 * @param binnaAshtakavarga - 2D array [planet][rasi] from BAV
 * @returns Reduced BAV after trikona sodhana
 */
export const trikonaSodhana = (binnaAshtakavarga: number[][]): number[][] => {
  // Deep copy
  const bav = binnaAshtakavarga.map(row => [...row]);

  // Process only planets 0-6 (Sun to Saturn), not Lagna
  for (let p = 0; p < 7; p++) {
    // There are 4 trine groups: (0,4,8), (1,5,9), (2,6,10), (3,7,11)
    for (let r = 0; r < 4; r++) {
      const r1 = r;
      const r2 = r + 4;
      const r3 = r + 8;

      const planetRow = bav[p];
      if (!planetRow) continue;

      const v1 = planetRow[r1] ?? 0;
      const v2 = planetRow[r2] ?? 0;
      const v3 = planetRow[r3] ?? 0;

      // Rule 1: If at least one rasi has zero, no reduction
      if (v1 === 0 || v2 === 0 || v3 === 0) {
        continue;
      }

      // Rule 2: If all three have the same value, make them all zero
      if (v1 === v2 && v2 === v3) {
        planetRow[r1] = 0;
        planetRow[r2] = 0;
        planetRow[r3] = 0;
      } else {
        // Rule 3: Subtract the minimum from all three
        const minValue = Math.min(v1, v2, v3);
        planetRow[r1] = (planetRow[r1] ?? 0) - minValue;
        planetRow[r2] = (planetRow[r2] ?? 0) - minValue;
        planetRow[r3] = (planetRow[r3] ?? 0) - minValue;
      }
    }
  }

  return bav;
};

/**
 * Ekadhipatya Sodhana - Dual lordship reduction
 *
 * Applies to Mars, Mercury, Jupiter, Venus, Saturn (planets 2-6)
 * who own two signs each.
 *
 * Rules:
 * 1. If either rasi has zero value, no reduction
 * 2. If both rasis are occupied by planets, no reduction
 * 3. If one is occupied and one empty:
 *    a. If empty rasi has lower value, make it zero
 *    b. If empty rasi has higher value, replace with occupied rasi's value
 * 4. If both rasis are empty:
 *    a. If same value, make both zero
 *    b. If different, replace higher with lower
 *
 * @param bavAfterTrikona - BAV after trikona sodhana
 * @param houseChart - Original house chart to check occupancy
 * @returns Reduced BAV after ekadhipatya sodhana (also called Sodhita Ashtakavarga)
 */
export const ekadhipatyaSodhana = (
  bavAfterTrikona: number[][],
  houseChart: HouseChart
): number[][] => {
  // Deep copy
  const bav = bavAfterTrikona.map(row => [...row]);

  // Rasi pairs owned by each planet (Mars to Saturn, indices 2-6)
  // Format: [rasi1, rasi2] where both are owned by the same planet
  const raasiOwners: [number, number][] = [
    [4, 3], // Mars owns Leo(4) and Cancer(3) - wait, that's wrong
    // Actually: Mars owns Aries(0) and Scorpio(7)
    // Let me check the Python code again...
  ];

  // From Python: rasi_owners=[4,3,(0,7),(2,5),(8,11),(1,6),(9,10)]
  // This means:
  // Index 0: Leo (4) - Sun's only sign
  // Index 1: Cancer (3) - Moon's only sign
  // Index 2: (0, 7) = Aries, Scorpio - Mars
  // Index 3: (2, 5) = Gemini, Virgo - Mercury
  // Index 4: (8, 11) = Sagittarius, Pisces - Jupiter
  // Index 5: (1, 6) = Taurus, Libra - Venus
  // Index 6: (9, 10) = Capricorn, Aquarius - Saturn

  // For ekadhipatya, we only process planets with dual ownership (Mars to Saturn)
  // Python loop: for p in range(2,7): r1,r2 = rasi_owners[p]
  // So p=2 -> rasi_owners[2] = (0,7)
  // p=3 -> rasi_owners[3] = (2,5)
  // etc.

  const dualOwnershipPairs: Record<number, [number, number]> = {
    2: [0, 7], // Mars: Aries, Scorpio
    3: [2, 5], // Mercury: Gemini, Virgo
    4: [8, 11], // Jupiter: Sagittarius, Pisces
    5: [1, 6], // Venus: Taurus, Libra
    6: [9, 10] // Saturn: Capricorn, Aquarius
  };

  // Check if a rasi is occupied (has any planet or Lagna in it)
  const isOccupied = (rasi: number): boolean => {
    const content = houseChart[rasi];
    return content !== undefined && content.trim() !== '';
  };

  for (let p = 2; p <= 6; p++) {
    const pair = dualOwnershipPairs[p];
    if (!pair) continue;
    const [r1, r2] = pair;
    const planetRow = bav[p];
    if (!planetRow) continue;

    const r1Occupied = isOccupied(r1);
    const r2Occupied = isOccupied(r2);

    const val1 = planetRow[r1] ?? 0;
    const val2 = planetRow[r2] ?? 0;

    // Rule 1: If either BAV value is 0, no reduction
    // Rule 2: If both rasis are occupied, no reduction
    if (val1 === 0 || val2 === 0) {
      continue;
    }
    if (r1Occupied && r2Occupied) {
      continue;
    }

    // Rule 4: Both rasis are empty
    if (!r1Occupied && !r2Occupied) {
      if (val1 === val2) {
        // Rule 4(a): Same value, make both zero
        planetRow[r1] = 0;
        planetRow[r2] = 0;
      } else {
        // Rule 4(b): Different values, replace higher with lower
        const minValue = Math.min(val1, val2);
        planetRow[r1] = minValue;
        planetRow[r2] = minValue;
      }
    } else {
      // Rule 3: One rasi is occupied, one is empty
      if (r1Occupied) {
        // r2 is empty
        if (val2 < val1) {
          // Rule 3(a): Empty rasi has lower value, make it zero
          planetRow[r2] = 0;
        } else {
          // Rule 3(b): Empty rasi has higher value, replace with occupied's value
          planetRow[r2] = val1;
        }
      } else {
        // r1 is empty
        if (val1 < val2) {
          // Rule 3(a): Empty rasi has lower value, make it zero
          planetRow[r1] = 0;
        } else {
          // Rule 3(b): Empty rasi has higher value, replace with occupied's value
          planetRow[r1] = val2;
        }
      }
    }
  }

  return bav;
};

/**
 * Calculate Sodhya Pindas from reduced Ashtakavarga
 *
 * @param bavAfterEkadhipatya - BAV after both trikona and ekadhipatya sodhana
 * @param houseChart - Original house chart for planet positions
 * @returns Rasi Pindas, Graha Pindas, and Sodhya Pindas for each planet (0-6)
 */
const calculateSodhyaPindas = (
  bavAfterEkadhipatya: number[][],
  houseChart: HouseChart
): SodhyaPindasResult => {
  const bav = bavAfterEkadhipatya;
  const pToH = getPlanetToHouseFromChart(houseChart);

  const raasiPindas: number[] = Array(7).fill(0);
  const grahaPindas: number[] = Array(7).fill(0);
  const sodhyaPindas: number[] = Array(7).fill(0);

  // Get planet houses for Sun to Saturn (0-6)
  const planetHouses: number[] = [];
  for (let p = 0; p < 7; p++) {
    planetHouses.push(pToH[p] ?? 0);
  }

  // Calculate Rasi Pindas: sum of (BAV[p][r] * rasimana_multiplier[r])
  for (let p = 0; p < 7; p++) {
    let sum = 0;
    for (let r = 0; r < 12; r++) {
      sum += bav[p][r] * RASIMANA_MULTIPLIERS[r];
    }
    raasiPindas[p] = sum;
  }

  // Calculate Graha Pindas: sum of (grahamana_multiplier[i] * BAV[p][house of planet i])
  for (let p = 0; p < 7; p++) {
    let sum = 0;
    for (let i = 0; i < 7; i++) {
      const planetRasi = planetHouses[i];
      sum += GRAHAMANA_MULTIPLIERS[i] * bav[p][planetRasi];
    }
    grahaPindas[p] = sum;
  }

  // Sodhya Pindas = Rasi Pindas + Graha Pindas
  for (let p = 0; p < 7; p++) {
    sodhyaPindas[p] = raasiPindas[p] + grahaPindas[p];
  }

  return { raasiPindas, grahaPindas, sodhyaPindas };
};

/**
 * Calculate Sodhya Pindas from Binna Ashtakavarga
 *
 * Applies Trikona Sodhana, then Ekadhipatya Sodhana, then calculates Pindas.
 *
 * @param binnaAshtakavarga - 2D array [planet][rasi] from getAshtakavarga
 * @param houseChart - Original house chart
 * @returns Rasi Pindas, Graha Pindas, and Sodhya Pindas
 */
export const sodhayaPindas = (
  binnaAshtakavarga: number[][],
  houseChart: HouseChart
): SodhyaPindasResult => {
  // Step 1: Trikona Sodhana
  const bavAfterTrikona = trikonaSodhana(binnaAshtakavarga);

  // Step 2: Ekadhipatya Sodhana (results in Sodhita Ashtakavarga)
  const bavAfterEkadhipatya = ekadhipatyaSodhana(bavAfterTrikona, houseChart);

  // Step 3: Calculate Pindas
  return calculateSodhyaPindas(bavAfterEkadhipatya, houseChart);
};

/**
 * Get complete Ashtakavarga analysis including Sodhya Pindas
 *
 * @param houseChart - House chart with planets
 * @returns Complete Ashtakavarga analysis
 */
export const getCompleteAshtakavarga = (
  houseChart: HouseChart
): AshtakavargaResult & { sodhyaPindas: SodhyaPindasResult } => {
  const avResult = getAshtakavarga(houseChart);
  const pindas = sodhayaPindas(avResult.binnaAshtakavarga, houseChart);

  return {
    ...avResult,
    sodhyaPindas: pindas
  };
};

/**
 * Get Kaksha (sub-division) of a planet for transit predictions
 *
 * Each rasi is divided into 8 kakshas of 3°45' each.
 * Kaksha lords are: Saturn, Jupiter, Mars, Sun, Venus, Mercury, Moon, Lagna
 *
 * @param longitude - Longitude in the sign (0-30)
 * @returns Kaksha index (0-7) and lord
 */
export const getKaksha = (
  longitude: number
): { kakshaIndex: number; kakshaLord: number | string } => {
  // Each kaksha spans 30/8 = 3.75 degrees
  const kakshaSize = 30 / 8;
  const kakshaIndex = Math.floor(longitude / kakshaSize) % 8;

  // Kaksha lords in order: Saturn(6), Jupiter(4), Mars(2), Sun(0),
  // Venus(5), Mercury(3), Moon(1), Lagna('L')
  const kakshaLords: (number | string)[] = [6, 4, 2, 0, 5, 3, 1, ASCENDANT_SYMBOL];

  return {
    kakshaIndex,
    kakshaLord: kakshaLords[kakshaIndex]
  };
};
